WIP: Android meson integration

rsglobal opened the pull request:
Require additional dependencies:

```
# Mainline libyaml with Android.mk file
git clone --branch android https://github.com/rsglobal/libyaml.git external/libyaml

# Chromium libyuv (as libyuv_chromium target)
git clone --branch android https://github.com/rsglobal/libyuv.git external/libyuv_chromium
```
How to use:

1. Enable libcamera, add pipelines and IPAs into your board.mk file:
```
board.mk:
    BOARD_LIBCAMERA_USES_MESON_BUILD := true
    BOARD_LIBCAMERA_IPAS := rkisp1
    BOARD_LIBCAMERA_PIPELINES := simple rkisp1
```

2. Add the following package configuration into your device.mk file:

```
device.mk:
    PRODUCT_PACKAGES += \
        camera.libcamera
    PRODUCT_PROPERTY_OVERRIDES += \
        ro.hardware.camera=libcamera
```

Example of `/vendor/etc/libcamera/camera_hal.yaml`:
https://chromium-review.googlesource.com/c/chromiumos/overlays/board-overlays/+/2866850/2/overlay-soraka-libcamera/media-libs/libcamera-configs/files/camera_hal.yaml

rsglobal commented:
CC: @johnstultz-work

rsglobal commented:
@kbingham ,

Hello Kieran,

Sorry for creating this PR in your personal repository. Hope you don't mind we do some preliminary review here until it will be ready for submission via mailing list.

kbingham reviewed (commented):
It's a bit unfortunate not to be able to use the standard library - but I guess that's a limitation on Android.
And given that this is an 'android hal' it's probably not unreasonable to highglight that as an issue.
We use the HAL in ChromeOS where we can utilise the std::filesystem.
I wonder if it can be wrapped in some way so we can take advantage of the std::filesystem when it's there or not otherwise.

Also you've droped a check, you're no longer validating that the file is there before opening it. I guess the difference is subtle as the fopen will likely fail anyway, but it will be a different error message reported. Looks like previously that only reports a Debug level, so maybe this is actually better...!?

kbingham commented on src/android/camera_hal_config.cpp:
```diff
@@ -357,15 +357,8 @@ CameraHalConfig::CameraHalConfig()
-	}
-
-	FILE *fh = fopen(filePath.c_str(), "r");
+	const char *filePath = LIBCAMERA_SYSCONF_DIR "camera_hal.yaml";
```
Does this correctly join the two paths? What happens if LIBCAMERA_SYSCONF_DIR doesn't have a / at the end?

rsglobal replied:
Nice catch. Fixed.

kbingham commented on src/android/camera_hal_config.cpp:
```diff
@@ -357,15 +357,8 @@ CameraHalConfig::CameraHalConfig()
  */
 int CameraHalConfig::parseConfigurationFile()
 {
-	std::filesystem::path filePath = LIBCAMERA_SYSCONF_DIR;
```
It's a bit unfortunate not to be able to use the standard library - but I guess that's a limitation on Android.
And given that this is an 'android hal' it's probably not unreasonable to highglight that as an issue.
We use the HAL in ChromeOS where we can utilise the std::filesystem.
I wonder if it can be wrapped in some way so we can take advantage of the std::filesystem when it's there or not otherwise.

rsglobal replied:
std::filesystem is present in AOSP but not available for vendor libraries.
To make it available `Android.bp` should specify `vendor: true` field.
But for some reason nobody added it there.

It is a single usage. Do we really want to have 2 implementations here depending on presence of std::filesystem?

kbingham commented on src/android/camera_hal_config.cpp:
```diff
@@ -357,15 +357,8 @@ CameraHalConfig::CameraHalConfig()
 {
-	std::filesystem::path filePath = LIBCAMERA_SYSCONF_DIR;
-	filePath /= "camera_hal.yaml";
-	if (!std::filesystem::is_regular_file(filePath)) {
```
Also you've dropped a check, you're no longer validating that the file is there before opening it. I guess the difference is subtle as the fopen will likely fail anyway, but it will be a different error message reported. Looks like previously that only reports a Debug level, so maybe this is actually better...!?

rsglobal replied:
So should I fstat() it and check is it a regular file or  better to keep as is?

kbingham commented on src/libcamera/meson.build:
```diff
@@ -71,8 +71,11 @@ subdir('pipeline')
-libgnutls = cc.find_library('gnutls', required : true)
+libdl = dependency('', required : false)
+if not cc.has_function('dlopen')
+    libdl = cc.find_library('dl')
```
does this need to be required : true  ?

kbingham replied:
Perhaps that's default. I can't recall.

rsglobal replied:
mesa3d meson.build was used as a reference for doing this.
I've created another var null_dep to make things more clear.

kbingham commented on src/libcamera/utils.cpp:
```diff
@@ -67,8 +67,8 @@ char *secure_getenv(const char *name)
 #else
-	if (issetugid())
-		return NULL;
+//	if (issetugid())
```
This should likely be wrapped better with a check that does "the right thing" if issetugid isn't present, and should have a define set for HAVE_SETUGID when it's available like the others.

It's fairly horrible code here anyway though (I mean before this patch)

rsglobal replied:
Modified to use HAVE_SETUGID

kbingham commented on Android.mk:
```diff
@@ -0,0 +1,75 @@
+# Mesa 3-D graphics library
```
John Stultz has been telling me about this other work! Sounds like some good progress on getting things moving.
I guess this is not mesa 3-d graphics library here though ;-)

rsglobal replied:
Yes. It was a copy-paste from mesa3d. I spent a day to make it buildable for libcamera and to get quick feedback. Changed.
Can I use Apache-2.0 license for Android.mk files?

kbingham commented on Android.mk:
```diff
@@ -0,0 +1,75 @@
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+# DEALINGS IN THE SOFTWARE.
+
+ifneq ($(filter true, $(BOARD_LIBCAMERA_USES_MESON_BUILD)),)
```
Is this needed? Isn't it always needed to be true?

rsglobal replied:
These Android.mks may exist out-of-tree, for example backported to earlier Android version.
Therefore we need a way of disabling in-tree targets.

kbingham commented on Android.mk:
```diff
@@ -0,0 +1,75 @@
+LOCAL_SHARED_LIBRARIES := $(__MY_SHARED_LIBRARIES)
+
+# Modules 'camera.libcamera', produces '/vendor/lib{64}/hw/camera.libcamera.so' HAL
+$(eval $(call libcamera-lib,camera.libcamera,hw,LIBCAMERA_BIN))
```
I'm planning to split out the android hal so it builds a separate .so rather than uses the libcamera.so - Just something to be aware of

rsglobal replied:
Sounds good. Probably `LOCAL_CHECK_ELF_FILES := false` can be dropped then. It fails at this moment because ELF `soname` is still `libcamera.so`, but filename is `camera.libcamera.so`.

kbingham commented on meson_cross.mk:
```diff
@@ -0,0 +1,247 @@
+# Mesa 3-D graphics library
```
Not Mesa ;-)

rsglobal replied:
Sure

kbingham commented on meson_cross.mk:
```diff
@@ -0,0 +1,247 @@
+	cd $(MESON_OUT_DIR) && PATH=/usr/bin:/usr/local/bin:$$PATH meson ./build \
+	--cross-file $(AOSP_ABSOLUTE_PATH)/$(MESON_GEN_DIR)/aosp_cross           \
+	--buildtype=release                                                      \
+	-Dandroid=enabled                                                        \
```
There is an android_platform option too, it defaults to 'generic' so I presume that is where we'll handle actual aosp specifics. 
May need something to handle things like any android specific gralloc handling perhaps or such.
Not needed here / now though

rsglobal replied:
Ok.

kbingham commented on meson_cross.mk:
```diff
@@ -0,0 +1,247 @@
+	-Dandroid=enabled                                                        \
+	-Dipas=$(subst $(space),$(comma),$(BOARD_LIBCAMERA_IPAS))                \
+	-Dpipelines=$(subst $(space),$(comma),$(BOARD_LIBCAMERA_PIPELINES))      \
+	-Dplatform-sdk-version=$(PLATFORM_SDK_VERSION)                           \
```
Where does platform-sdk-version come from?

rsglobal replied:
It's a leftover from mesa3d. Removed.

kbingham commented:
If we can resolve the workaround/hacks, I'd say the meson_cross.mk should be something we could integrate into libcamera! Very interesting work.
Needs posting through the libcamera mailing list of course though.

rsglobal commented:
Sounds good.

Do you know if raspberry pi guys have any plans to drop boost dependency?

kbingham commented:
> Sounds good.
> 
> Do you know if raspberry pi guys have any plans to drop boost dependency?

It's used for json parsing and property trees. We're trying to find a way to not need boost but it hasn't been a high priority yet.
You can disable the boost requirement by not enabling the RPi IPA.

kbingham commented:
Just been takling to Laurent about this - he's concerned that the android.mk files support will be deprecated and replaced by .bp files, which are only declarative. Is there a plan to manage this with Mesa?

rsglobal commented:
> Just been takling to Laurent about this - he's concerned that the android.mk files support will be deprecated and replaced by .bp files, which are only declarative. Is there a plan to manage this with Mesa?

We know that someday Google plan to stop supporting Android.mk's .
Also Google announced that it will switch to Bazel as a replacement of Soong.
But since date for both is unknown, and Android.mk is the only option known ATM, we have no plans to manage this.

rsglobal commented:
@pinchartl,

I just saw https://github.com/kbingham/libcamera/commit/969da31894394db4f64c4a0e96ec2252fb13142b which is causing conflicts. It is very similar to my ("libcamera: Make issetugid() optional").

I can assume someone else is also working on Android integration.
Otherwise please don't hesitate to at least keep `Suggested-by` tag. Or push me to up-stream it next time.

kbingham commented:
Hi @rsglobal 

There is another team which has been working on android for quite some time as well.
Though they have taken a different approach to you, and they are manually creating blue prints, so I don't think they overlap with the same goals.

pinchartl commented:
Commit 969da31 has been developed independently from this pull request. I'm sorry for not having noticed your patch first, the change came from a hack I noticed in a parallel development, as mentioned by Kieran. I'm also afraid I can't add a Suggested-by tag as the commit has been merged already.

I'd recommend submitting patches to the libcamera-devel mailing list, as that's what we use for development. Kieran is kind enough to follow issues and pull requests on github, but he can't scale by himself :-)

pinchartl commented on src/libcamera/meson.build:
```diff
@@ -70,9 +70,14 @@ subdir('ipa')
+if not cc.has_function('dlopen')
+    libdl = cc.find_library('dl')
+endif
+libgnutls = cc.find_library('gnutls', required : false)
```
This changed from required : true to required : false. If that was intentional, it should be moved to a separate commit, with an explanation in the commit message.

rsglobal commented:
Updated:
* Fixed missing libcamea.so and libcamera-base.so deps
* Small fixes in makefiles
* First successful run. I can see the logs probing for camera device.

rsglobal commented:
Latest new:
Looks like meson build within AOSP script can build shared libraries only, but can't build executables.
Executables require different set of linker arguments, so it require additional handling.
And libcamera uses proxy executable for IPA (is it really necessary?....). So I'm currently stuck on bringing-up camera for rpi4.

rsglobal commented:
@kbingham 

Looks like raspberrypi pipeline is incomplete. It doesn't set bufferscount during validation as ipu3 does, which disapoints the Android.

kbingham commented:
Interesting, the RPi hasn't been tested much with the Android layer, but I'd like to see it tested more.
I'd need either a working Chrome or AOSP image for that though.

I see that there was some work to refactor bufferCount recently, but it has stalled. The latest version is https://lists.libcamera.org/pipermail/libcamera-devel/2021-August/023773.html

This might be a path towards solving this - or otherwise, initialising a default bufferCount of say 4 in the RPi pipeline handler might help.

Can you post this on the mailing list perhaps?

kbingham commented:
I keep meaning to try: https://github.com/FydeOS/chromium_os-raspberry_pi/releases/tag/r92 but I haven't had any opportunity or time to do so.

kbingham commented:
> Sounds good.
> 
> Do you know if raspberry pi guys have any plans to drop boost dependency?

This is dropped now by the way. I wonder if it helps any development here.

rsglobal commented:
> 
> This is dropped now by the way. I wonder if it helps any development here.

I'll check it soon, thanks

rothn commented:
Hey! @rsglobal @kbingham , I'm interested in this too and can offer help, a second pair of eyes, etc. if that would help get this merged to git.libcamera.org. Let me know!

Additional context: https://github.com/waydroid/waydroid/issues/519

kbingham commented:
Can you rebase to the latest libcamera, and make sure everything compiles successfully, then post to the libcamera mailing list please?

rsglobal commented:
> > This is dropped now by the way. I wonder if it helps any development here.
> 
> I'll check it soon, thanks

It works.

> Can you rebase to the latest libcamera, and make sure everything compiles successfully, then post to the libcamera mailing list please?

Yes, but it would be nice to have code owner for these files from libcamera side. Otherwise it won't work for a long period.

pinchartl commented:
The first two commits can be reviewed on the list and merged. The third one is more problematic, as make-based build are deprecated in Android (see https://source.android.com/docs/setup/build).

rsglobal commented:
> The first two commits can be reviewed on the list and merged. The third one is more problematic, as make-based build are deprecated in Android (see https://source.android.com/docs/setup/build).

It was deprecated long time ago nevertheless it is still supported and currently works and used by mainline mesa3d and everybody is happy :) . Building using NDK is an alternative, but it may have a different set of challenges, and NDK doesn't allow to build it within AOSP.

rothn commented:
@rsglobal, My understanding is that @pinchartl's comment refers to Android's new [Soong build system](https://source.android.com/docs/setup/build), based on Bazel. So, Android.mk would be replaced by Android.bp.

The main challenge here is that libcamera uses the Soong build system. We have at least a few options if we want to use an Android.bp instead of the deprecated Android.mk:
* Transition libcamera to bazel and build intuitively (best, simplest, most consistent, but potentially disruptive)
* Invoke Soong from bazel with a [genrule](https://bazel.build/reference/be/general#genrule), akin to what meson_cross.mk does (complex-- I would be happy to do this but I'll need help understanding how meson_cross.mk works; probably harder than writing meson_cross.mk was)

If there's appetite for either of these options, I'm happy to help out where I can. I use Bazel regularly at work, but I'll need help understanding these Makefiles.

For example, we'd have something like this:
```bzl
cc_library_shared {
    name: "camera.libcamera",
    relative_install_path: "hw",
    vendor: true,
    srcs: [
        "foo",
        "bar",
    ],
    shared_libs: [
        "libc",
        "libexif",
        "libjpeg",
        "libyuv_chromium",
        "libdl",
        "libyaml",
    ],
}

cc_library_shared {
    name: "libcamera",
    vendor: true,
    srcs: [
        "foo",
        "bar",
    ],
    shared_libs: [
        "libc",
        "libexif",
        "libjpeg",
        "libyuv_chromium",
        "libdl",
        "libyaml",
    ],
}

cc_library_shared {
    name: "libcamera-base",
    vendor: true,
    srcs: [
        "foo",
        "bar",
    ],
    shared_libs: [
        "libc",
        "libexif",
        "libjpeg",
        "libyuv_chromium",
        "libdl",
        "libyaml",
    ],
}
```

...Instead of this:
```make
# SPDX-License-Identifier: Apache-2.0
#
# Copyright (C) 2021, GlobalLogic Ukraine
# Copyright (C) 2021, Roman Stratiienko (r.stratiienko@gmail.com)
#
# Android.mk - Android makefile
#

ifneq ($(filter true, $(BOARD_LIBCAMERA_USES_MESON_BUILD)),)

LOCAL_PATH := $(call my-dir)
LIBCAMERA_TOP := $(dir $(LOCAL_PATH))

include $(CLEAR_VARS)

LOCAL_SHARED_LIBRARIES := libc libexif libjpeg libyuv_chromium libdl libyaml
MESON_GEN_PKGCONFIGS := libexif libjpeg yaml-0.1 libyuv dl

ifeq ($(TARGET_IS_64_BIT),true)
LOCAL_MULTILIB := 64
else
LOCAL_MULTILIB := 32
endif
include $(LOCAL_PATH)/meson_cross.mk

ifdef TARGET_2ND_ARCH
LOCAL_MULTILIB := 32
include $(LOCAL_PATH)/meson_cross.mk
endif

#-------------------------------------------------------------------------------

define libcamera-lib
LOCAL_MODULE_CLASS := SHARED_LIBRARIES
LOCAL_MODULE := $1
LOCAL_VENDOR_MODULE := true
LOCAL_MODULE_RELATIVE_PATH := $2
ifdef TARGET_2ND_ARCH
LOCAL_SRC_FILES_$(TARGET_ARCH) := $(call relative_top_path,$(LOCAL_PATH))$($3)
LOCAL_SRC_FILES_$(TARGET_2ND_ARCH) := $(call relative_top_path,$(LOCAL_PATH))$(2ND_$3)
LOCAL_MULTILIB := both
else
LOCAL_SRC_FILES := $(call relative_top_path,$(LOCAL_PATH))$($3)
endif
LOCAL_CHECK_ELF_FILES := false
LOCAL_MODULE_SUFFIX := .so
include $(BUILD_PREBUILT)
include $(CLEAR_VARS)
endef

__MY_SHARED_LIBRARIES := $(LOCAL_SHARED_LIBRARIES)
include $(CLEAR_VARS)
LOCAL_SHARED_LIBRARIES := $(__MY_SHARED_LIBRARIES)

# Modules 'libcamera', produces '/vendor/lib{64}/libcamera.so'
$(eval $(call libcamera-lib,libcamera,,LIBCAMERA_BIN))
# Modules 'libcamera-base', produces '/vendor/lib{64}/libcamera-base.so'
$(eval $(call libcamera-lib,libcamera-base,,LIBCAMERA_BASE_BIN))

LOCAL_SHARED_LIBRARIES += libcamera libcamera-base
# Modules 'camera.libcamera', produces '/vendor/lib{64}/hw/camera.libcamera.so' HAL
$(eval $(call libcamera-lib,camera.libcamera,hw,LIBCAMERA_HAL_BIN))

#-------------------------------------------------------------------------------

endif
```

rothn commented:
I'll add that a lot of thought has been put into steering people away from using `genrule()`s like the one described, so I fully expect this to be more challenging than writing meson_cross.mk if not completely infeasible. By far the best option would seem to be moving libcamera to Bazel if we need a .bp, or accepting a deprecated .mk if the libcamera team doesn't want to use Bazel. Considering that 2/3 use-cases (ChromeOS, Android) for libcamera are Google ecosystems, Bazel might be the best way to go.

Interested to read your thoughts on this too, @kbingham!

rsglobal commented:
@rothn ,

Each solution has its pros and cons, maintainers should decide which way to go.  

@kbingham @pinchartl ,

Since libcamera advertises Android support , could you tell us more how and on what hardware it's used now? It would be very nice to have some android-related how-to docs.

kbingham commented:
I would very much like it to be easier to use libcamera with AOSP, so this deserves a discussion on the libcamera mailing list (https://lists.libcamera.org/listinfo/libcamera-devel) not here.

I can't see us changing the build system drastically though, we've chosen meson already - so we'd need to see real clear benefits for the change.

We already support ChromeOS with meson.

Keeping two builds in parallel will be a pain as well though, but if a build for android exists, I think perhaps it's better to live in libcamera.git than outside. At least then it's centralised - but we'd probably want a way to easily automate building against AOSP/NDK so we can track it.

It seems I have a lot of questions about how we can handle this - but I'd really prefer we discuss the options on the actual development list so others can be involved, not just me.

kbingham commented:
If we merge the Android.mk file, how do we build test it?

kbingham commented:
> Since libcamera advertises Android support , could you tell us more how and on what hardware it's used now? It would be very nice to have some android-related how-to docs.

The Android HAL is used by ChromeOS. We currently use this with Intel IPU3 Chromebooks, and I believe it can be used already by a Rockchip RK3399 ChromeTab too, though I don't think that one has been used in quite a while. Most of our development has been with the IPU3.

rsglobal commented:
> @rsglobal Is this something that you are willing to do ? I.e. will you post the support to the libcamera mailing list, and provide documentation on how to build/use it ?

Unfortunately I don't have enough free time for that right now. Also @pinchartl statements that Android.mk is obsolete makes me feel that I'm risking to waste a lot of time without eventually get things up-streamed.

kbingham commented:
I'm happy to merge an Android.mk if we can build test it to catch breakage. If Android.mk options become obsolete by Android, then it won't work past that version, but it would still work for earlier versions that the deprecation - so I still think there's value.

But ... the patches need to be posted to the libcamera-devel mailing list - They won't get merged here.

kbingham commented:
We're a small team and resource constrained. I don't think we have anyone that could do that right now.

@rothn , as you resumed this conversation here, is this something you want to take on?

rothn commented:
Yes, I'm happy to help, not least because this happens to support my
phone's camera :). I'd like to take a stab at creating a parallel
Android.bp which would be continuously tested with "blaze test ..." or
similar. Let me know if that would work for you.

I'll try posting this to the mailing list as well later this evening along
with some context, and you may prefer to reply there.

On Tue, Oct 11, 2022, 1:10 PM Kieran Bingham ***@***.***>
wrote:

> We're a small team and resource constrained. I don't think we have anyone
> that could do that right now.
>
> @rothn <https://github.com/rothn> , as you resumed this conversation
> here, is this something you want to take on?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kbingham/libcamera/pull/26#issuecomment-1275087711>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABDTDN6LB3OGYQN5PPINQD3WCWUSJANCNFSM46J7WCKA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

rsglobal commented:
> Yes, I'm happy to help, not least because this happens to support my
phone's camera :). I'd like to take a stab at creating a parallel
Android.bp

Mesa3d was living with similar approach for a long time. Once a week it required tuning and therefore Android build was usually broken. With current Android.mk approach we have just a 1 fix per 6 month which is a way better.

Building using android NDK should similar to building for pure linux.

rsglobal commented:
https://docs.mesa3d.org/android.html

kbingham commented:
> https://docs.mesa3d.org/android.html

Thanks, if we merge support to build android in libcamera, then some equivalent documentation for this should be added to libcamera too so it can be referenced for how people can use it.

kbingham commented:
> > Yes, I'm happy to help, not least because this happens to support my
> > phone's camera :). I'd like to take a stab at creating a parallel
> > Android.bp
> 
> Mesa3d was living with similar approach for a long time. Once a week it required tuning and therefore Android build was usually broken. With current Android.mk approach we have just a 1 fix per 6 month which is a way better.

Wrapping the existing meson build sounds more supportable than a parallel build that needs constant maintenance everytime we make a change / add a new object  / compile flags etc...
But lets follow that discussion on the list.

> Building using android NDK should similar to building for pure linux.

Good, that makes it easy to incorporate into the build tests.
